### ISSUE TITLE
XL: With insufficient write quorum, put object still succeeds 

### ISSUE THREAD

**Issue by balamurugana** [minio/minio#1951]
In a 8 disks setup, where three disks are down, put object still succeeds.   As per write quorum calculation, the setup can withstand max of no.of disks - no. of disks / 2 + 2 ie 2 disk failure.

**Comment by harshavardhana** [comment]
Not able to reproduce this with current master

terminal 1 

```
ERRO[0485] Unable to initiate new multipart upload id.   cause=disk not found sysInfo=map[host.arch:amd64 host.name:command.local host.cpus:4 mem.used:2.6MB host.os:darwin host.lang:go1.6 mem.total:43MB mem.heap.used:2.6MB mem.heap.total:38MB] type=*errors.errorString
ERRO[0485] Unable to initiate new multipart upload id.   cause=disk not found sysInfo=map[host.arch:amd64 host.lang:go1.6 host.cpus:4 mem.used:2.6MB mem.total:43MB mem.heap.total:38MB host.os:darwin host.name:command.local mem.heap.used:2.6MB] type=*errors.errorString
ERRO[0486] Unable to initiate new multipart upload id.   cause=disk not found sysInfo=map[host.arch:amd64 host.name:command.local host.lang:go1.6 mem.used:2.6MB host.os:darwin host.cpus:4 mem.total:43MB mem.heap.used:2.6MB mem.heap.total:38MB] type=*errors.errorString
ERRO[0487] Unable to initiate new multipart upload id.   cause=disk not found sysInfo=map[mem.heap.total:38MB mem.used:2.7MB mem.total:43MB mem.heap.used:2.7MB host.lang:go1.6 host.cpus:4 host.os:darwin host.arch:amd64 host.name:command.local] type=*errors.errorString
ERRO[0488] Unable to initiate new multipart upload id.   cause=disk not found sysInfo=map[mem.heap.total:38MB host.os:darwin host.name:command.local host.lang:go1.6 mem.total:43MB mem.heap.used:2.7MB host.arch:amd64 host.cpus:4 mem.used:2.7MB] type=*errors.errorString
```

terminal 2 

```
$ mc cp -r /usr/bin test/testbucket
mc: <ERROR> Unable to prepare URL for copying. Requested file ‘/usr/bin/vagrant’ has broken symlink
mc: <ERROR> Failed to copy ‘/usr/bin/emacs’. 500 Internal Server Error
mc: <ERROR> Session safely terminated. To resume session ‘mc session resume kEcAvekD’
```

terminal 3

```
$ mv /tmp/bc1 /tmp/bc1.dead & mv /tmp/bc2 /tmp/bc2.dead & mv /tmp/bc3 /tmp/bc3.dead[1] 50418
[2] 50419
[1]-  Done                    mv /tmp/bc1 /tmp/bc1.dead
[2]+  Done                    mv /tmp/bc2 /tmp/bc2.dead
```

Now trying to resume the operation 

```
$ mc session resume kEcAvekD
mc: <ERROR> Failed to copy ‘/usr/bin/emacs’. 500 Internal Server Error
mc: <ERROR> Session safely terminated. To resume session ‘mc session resume kEcAvekD’
```

Now with debug we can see a wrong error being generated but PutObject never completes or resumes.

```
mc: <DEBUG> POST /testbucket/bin/emacs?uploads= HTTP/1.1
Host: 127.0.0.1:9001
User-Agent: Minio (darwin; amd64) minio-go/1.0.1 mc/DEVELOPMENT.GOGET
Content-Length: 0
Authorization: AWS4-HMAC-SHA256 Credential=**REDACTED**/20160626/us-east-1/s3/aws4_request, SignedHeaders=expect;host;x-amz-content-sha256;x-amz-date, Signature=**REDACTED**
Content-Type: application/octet-stream
Expect: 100-continue
X-Amz-Content-Sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
X-Amz-Date: 20160626T073245Z
Accept-Encoding: gzip

mc: <DEBUG> HTTP/1.1 500 Internal Server Error
Transfer-Encoding: chunked
Accept-Ranges: bytes
Content-Type: text/xml; charset=utf-8
Date: Sun, 26 Jun 2016 07:32:45 GMT
Server: Minio/DEVELOPMENT.GOGET (darwin; amd64)
Vary: Origin
X-Amz-Request-Id: P0GC47LNOHTD9612

116
<?xml version="1.0" encoding="UTF-8"?>
<Error><Code>InternalError</Code><Message>We encountered an internal error, please try again.</Message><Key></Key><BucketName></BucketName><Resource>/testbucket/bin/emacs</Resource><RequestId>3L137</RequestId><HostId>3L137</HostId></Error>
0

mc: <DEBUG> Response Time:  4.85517ms

mc: <ERROR> Failed to copy ‘/usr/bin/emacs’. 500 Internal Server Error
 (3) cp-main.go:346 main.doCopySession.func1(..) Tags: [/usr/bin/emacs]
 (2) cp-main.go:207 main.doCopy(..) Tags: [http://127.0.0.1:9001/testbucket/bin/emacs]
 (1) common-methods.go:85 main.putTargetStreamFromAlias(..) Tags: [test, http://127.0.0.1:9001/testbucket/bin/emacs]
 (0) client-s3.go:252 main.(*s3Client).Put(..)
 Release-Tag:DEVELOPMENT.GOGET | Commit:DEVELOPMENT. | Host:command.local | OS:darwin | Arch:amd64 | Lang:go1.6 | Mem:1.9MB/8.2MB | Heap:1.9MB/4.7MB
mc: <ERROR> Session safely terminated. To resume session ‘mc session resume kEcAvekD’
```

**Comment by harshavardhana** [comment]
Was able to reproduce.  Sent a patch with the fix at #1986, additionally adds unit tests to catch quorum issue.
